src/main/java/jenkins/plugins/git/maintenance/TaskExecutor.java

```diff
@@ -0,0 +1,182 @@
+package jenkins.plugins.git.maintenance;
+
+import hudson.FilePath;
+import hudson.model.TaskListener;
+import hudson.plugins.git.GitTool;
+import hudson.plugins.git.util.GitUtils;
+import hudson.util.LogTaskListener;
+import jenkins.model.Jenkins;
+import org.jenkinsci.plugins.gitclient.CliGitAPIImpl;
+import org.jenkinsci.plugins.gitclient.Git;
+import org.jenkinsci.plugins.gitclient.GitClient;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class TaskExecutor implements Runnable {
+
+    private volatile boolean isThreadAlive;
+    Task maintenanceTask;
+    private List<GitMaintenanceSCM.Cache> caches;
+
+    private static final Logger LOGGER = Logger.getLogger(TaskExecutor.class.getName());
+
+    public TaskExecutor(Task maintenanceTask){
+        this.maintenanceTask = new Task(maintenanceTask);
+        caches = getCaches();
+        isThreadAlive = true;
+        LOGGER.log(Level.FINE,"New Thread created to execute " + maintenanceTask.getTaskName());
+    }
+
+    @Override
+    public void run() {
+
+        LOGGER.log(Level.FINE,"Executing maintenance task " + maintenanceTask.getTaskName() + " on git caches.");
+        GitClient gitClient;
+        TaskType taskType = maintenanceTask.getTaskType();
+        try {
+            for (GitMaintenanceSCM.Cache cache : caches) {
+                if (isThreadAlive) {
+                    // For now adding lock to all kinds of maintenance tasks. Need to study on which task needs a lock and which doesn't.
+                    Lock lock = cache.getLock();
+                    File cacheFile = cache.getCacheFile();
+                    try {
+                        gitClient = getGitClient(cacheFile);
+                        if (gitClient == null)
+                            return;
+
+                        lock.lock();
+                        LOGGER.log(Level.FINE, "Cache " + cacheFile.getName() + " locked.");
+                        executeMaintenanceTask(gitClient, taskType);
```

Comment:
I have some concerns over this locking mechanism.

- running a git maintenance task can be time intensive, locking a thread while others are waiting to acquire the lock might increase the number of waiting threads (need to take a jvm thread dump to analyse and confirm this)
- we should try to reduce the scope of locking instead of locking the whole git maintenance process, i.e, identify the operations that need locking and only apply it conditionally
- another strategy to use locks, lock.tryLock(time interval). This allows the waiting threads to wait for a certain time and move on if the subjected lock to be acquired is not free.